src/backend/catalog/storage_directory_table.c

```diff
@@ -28,45 +29,75 @@
 #include "cdb/cdbvars.h"
 
 /*
- * TODO: Redo pending delete
+ * TODO: support ufile pending delete xlog
+ *
+ * Ufile do not support deleteing files during WAL redo, two of reason:
+ *
+ * 1. deleting files requires a connection to object storage system.
+ * In order to establish the connection to the object storage, we
+ * need to access the catalog table to retrieve the connection
+ * configuration info, which is impossible during WAL redo.
+ *
+ * 2. no custom xlog entry support.
+ * Custom WAL Resource Managers are immature and not reflected in CBDB.
  *
- * We do not support deleteing files during WAL redo, this is because deleting
- * files requires a connection to object storage system. In order to establish
- * the connection to the object storage, we need to access the catalog table to
- * retrieve the connection configuration info, which is impossible during WAL
- * redo.
  */
-
 typedef struct UFileNodePendingDelete
 {
-	char  relkind;
-	Oid   spcId;			/* directory table needs an extra tabpespace */
-	char *relativePath;
-} UFileNodePendingDelete;
+	char		relkind;
+	Oid			spcId;			/* directory table needs an extra tabpespace */
+	char	   *relativePath;
+}			UFileNodePendingDelete;
 
 typedef struct PendingRelDeleteUFile
 {
-	UFileNodePendingDelete filenode;		/* relation that may need to be deleted */
-	bool		atCommit;		/* T=delete at commit; F=delete at abort */
-	int			nestLevel;		/* xact nesting level of request */
-	struct PendingRelDeleteUFile *next;		/* linked-list link */
-} PendingRelDeleteUFile;
+	PendingRelDelete reldelete; /* base pending delete */
+	UFileNodePendingDelete filenode;	/* relation that may need to be
+										 * deleted */
+}			PendingRelDeleteUFile;
+
+
+static void
+UfileDestroyPendingRelDelete(PendingRelDelete *reldelete)
+{
+	PendingRelDeleteUFile *ufiledelete;
+
+	Assert(reldelete);
+	ufiledelete = (PendingRelDeleteUFile *) reldelete;
+
+	pfree(ufiledelete->filenode.relativePath);
+	pfree(ufiledelete);
+}
+
+static void
+UfileDoPendingRelDelete(PendingRelDelete *reldelete)
+{
+	PendingRelDeleteUFile *ufiledelete;
 
-static PendingRelDeleteUFile *pendingDeleteUFiles = NULL; /* head of linked list */
+	Assert(reldelete);
+	ufiledelete = (PendingRelDeleteUFile *) reldelete;
+
+	UFileUnlink(ufiledelete->filenode.spcId, ufiledelete->filenode.relativePath);
```

Comment:
According to wenchao's review comments, need to change here.

Reply:
no here, i think it's L133, let us discuss at morning meeting.

Reply:
changed at last commit. PTAL